File: packages/sdk-python/tracelane/instrumentations/openai_agents.py
"""OpenAI Agents SDK instrumentation for Tracelane.

Wraps the openai.agents Runner class to emit OTel spans for every agent
run. Captures agent name, final output length, and run metadata.
The OpenAI Agents SDK (formerly Swarm) is a common agent runtime; this
adapter instruments it without capturing keys or message content.

Example::

    from openai.agents import Agent, Runner
    from tracelane.instrumentations.openai_agents import instrument_openai_agents

    instrument_openai_agents(Runner)
    agent = Agent(name="my-agent", instructions="...")
    result = await Runner.run(agent, "hello")
    # Span emitted for the full agent run
"""

from __future__ import annotations

from typing import Any

from opentelemetry import trace
from opentelemetry.trace import SpanKind, StatusCode

_tracer = trace.get_tracer("tracelane.openai-agents", "0.1.0")


def instrument_openai_agents(runner_cls: Any) -> None:
    """Instrument the OpenAI Agents Runner class to emit OTel spans.

    Patches both Runner.run (async) and Runner.run_sync (sync) class methods.

    Args:
        runner_cls: The openai.agents.Runner class (not an instance).

    Note:
        The agent name (agent.name) is captured as ai.agent.name.
        Handoff chains are tracked via ai.agent.handoff_count.
    """
    _patch_run(runner_cls)
    _patch_run_sync(runner_cls)


def _patch_run(runner_cls: Any) -> None:
    if not hasattr(runner_cls, "run"):
        return
    original = runner_cls.run

    async def _patched_run(agent: Any, input: Any, *args: Any, **kwargs: Any) -> Any:
        agent_name = getattr(agent, "name", "unknown")
        with _tracer.start_as_current_span(
            "openai_agents.runner.run",
            kind=SpanKind.CLIENT,
            attributes={
                "gen_ai.provider.name": "openai-agents",
                "ai.agent.name": agent_name,
                "openai_agents.input_length": len(str(input)) if input else 0,
            },
        ) as span:
            try:
                result = await original(agent, input, *args, **kwargs)
                _record_result(span, result)
                span.set_status(StatusCode.OK)
                return result
            except Exception as exc:
                span.record_exception(exc)
                span.set_status(StatusCode.ERROR, str(exc))
                raise

    runner_cls.run = _patched_run


def _patch_run_sync(runner_cls: Any) -> None:
    if not hasattr(runner_cls, "run_sync"):
        return
    original = runner_cls.run_sync

    def _patched_run_sync(agent: Any, input: Any, *args: Any, **kwargs: Any) -> Any:
        agent_name = getattr(agent, "name", "unknown")
        with _tracer.start_as_current_span(
            "openai_agents.runner.run_sync",
            kind=SpanKind.CLIENT,
            attributes={
                "gen_ai.provider.name": "openai-agents",
                "ai.agent.name": agent_name,
                "openai_agents.input_length": len(str(input)) if input else 0,
            },
        ) as span:
            try:
                result = original(agent, input, *args, **kwargs)
                _record_result(span, result)
                span.set_status(StatusCode.OK)
                return result
            except Exception as exc:
                span.record_exception(exc)
                span.set_status(StatusCode.ERROR, str(exc))
                raise

    runner_cls.run_sync = _patched_run_sync


def _record_result(span: Any, result: Any) -> None:
    final_output = getattr(result, "final_output", None)
    if final_output is not None:
        span.set_attribute("openai_agents.output_length", len(str(final_output)))
    messages = getattr(result, "messages", None) or getattr(result, "new_messages", None)
    if isinstance(messages, list):
        span.set_attribute("openai_agents.messages_count", len(messages))
    # Handoff chains: RunResult.next_agent or multi-turn results
    next_agent = getattr(result, "next_agent", None)
    if next_agent is not None:
        span.set_attribute("ai.agent.handoff_count", 1)
        next_name = getattr(next_agent, "name", "unknown")
        span.set_attribute("ai.agent.handoff_target", next_name)

File: packages/sdk-typescript/src/instrumentations/openai_agents.ts
/**
 * OpenAI Agents SDK instrumentation for Tracelane.
 *
 * Wraps the OpenAI Agents SDK Runner.run() class method to emit OTel spans
 * for every agent execution. Captures agent name, output length, and
 * handoff metadata. The Agents SDK is the direct route to OpenAI Frontier
 *
 * @example
 * ```ts
 * import { Agent, Runner } from "openai/agents";
 * import { instrumentOpenAIAgents } from "@tracelanedev/sdk/openai_agents";
 *
 * instrumentOpenAIAgents(Runner);
 * const result = await Runner.run(agent, "hello");
 * ```
 */

import { SpanKind, SpanStatusCode, trace } from "@opentelemetry/api";

const tracer = trace.getTracer("@tracelanedev/sdk-openai-agents", "0.1.0");

interface RunnerLike {
	run: (...args: unknown[]) => Promise<unknown>;
	runSync?: (...args: unknown[]) => unknown;
}

/**
 * Instrument the OpenAI Agents Runner class to emit OTel spans.
 *
 * @param runnerClass - The openai/agents Runner class (not an instance)
 */
export function instrumentOpenAIAgents(runnerClass: RunnerLike): void {
	_patchRun(runnerClass);
	if (runnerClass.runSync) {
		_patchRunSync(
			runnerClass as RunnerLike & { runSync: (...args: unknown[]) => unknown },
		);
	}
}

function _patchRun(runnerClass: RunnerLike): void {
	const originalRun = runnerClass.run.bind(runnerClass);

	runnerClass.run = async (...args: unknown[]) => {
		const agent = args[0] as Record<string, unknown> | undefined;
		const agentName = typeof agent?.name === "string" ? agent.name : "unknown";
		const input = args[1];

		return tracer.startActiveSpan(
			"openai_agents.runner.run",
			{
				kind: SpanKind.CLIENT,
				attributes: {
					"gen_ai.provider.name": "openai-agents",
					"ai.agent.name": agentName,
					"openai_agents.input_length":
						typeof input === "string" ? input.length : 0,
				},
			},
			async (span) => {
				try {
					const result = (await originalRun(...args)) as Record<
						string,
						unknown
					>;
					_recordRunResult(span, result);
					span.setStatus({ code: SpanStatusCode.OK });
					return result;
				} catch (e) {
					span.recordException(e as Error);
					span.setStatus({ code: SpanStatusCode.ERROR, message: String(e) });
					throw e;
				} finally {
					span.end();
				}
			},
		);
	};
}

function _patchRunSync(
	runnerClass: RunnerLike & { runSync: NonNullable<RunnerLike["runSync"]> },
): void {
	const originalRunSync = runnerClass.runSync.bind(runnerClass);

	runnerClass.runSync = (...args: unknown[]) => {
		const agent = args[0] as Record<string, unknown> | undefined;
		const agentName = typeof agent?.name === "string" ? agent.name : "unknown";
		const span = tracer.startSpan("openai_agents.runner.run_sync", {
			kind: SpanKind.CLIENT,
			attributes: {
				"gen_ai.provider.name": "openai-agents",
				"ai.agent.name": agentName,
			},
		});
		try {
			const result = originalRunSync(...args) as Record<string, unknown>;
			_recordRunResult(span, result);
			span.setStatus({ code: SpanStatusCode.OK });
			return result;
		} catch (e) {
			span.recordException(e as Error);
			span.setStatus({ code: SpanStatusCode.ERROR, message: String(e) });
			throw e;
		} finally {
			span.end();
		}
	};
}

function _recordRunResult(
	span: ReturnType<typeof tracer.startSpan>,
	result: Record<string, unknown>,
): void {
	const finalOutput = result.final_output ?? result.finalOutput;
	if (finalOutput !== undefined && finalOutput !== null) {
		span.setAttribute(
			"openai_agents.output_length",
			String(finalOutput).length,
		);
	}
	const messages = result.messages ?? result.new_messages;
	if (Array.isArray(messages)) {
		span.setAttribute("openai_agents.messages_count", messages.length);
	}
	const nextAgent = result.next_agent ?? result.nextAgent;
	if (nextAgent !== null && nextAgent !== undefined) {
		const nextName = (nextAgent as Record<string, unknown>).name;
		span.setAttribute("ai.agent.handoff_count", 1);
		if (typeof nextName === "string") {
			span.setAttribute("ai.agent.handoff_target", nextName);
		}
	}
}
